### ISSUE TITLE
Server crash when pasting

### ISSUE THREAD

**Issue by Postremus** [SBPrime/AsyncWorldEdit-Premium#11]
We updated our Server today to  the premium plugin. We randomly get this dump now.
https://gist.github.com/Postremus/df2683790349b5e68e6b

We are pasting the schematics with following code:

```
        try {
            EditSession es = we.getEditSessionFactory().getEditSession(BukkitUtil.getLocalWorld(location.getWorld()), 32000000);
            CuboidClipboard cc = MCEditSchematicFormat.MCEDIT.load(schematicFile);
            cc.paste(es, BukkitUtil.toVector(location));
        } catch (Exception ex) {
            ex.printStackTrace();
        }
```

**Comment by SBPrime** [comment]
There is 1 big difference between AWE and AWE premium. When the free version detects that the global queue is full it drops the blocks. When the premium version detects that the queue is full it pauses the thread that is adding the blocks. I assume that in your situation youdo the pasting from the main thread and this results in main thread lock.

There are few solutions to this:
1. Disable the block limit in AWE config
2. Implement the paste using the AWE API (I'll help provide support for this)
3. Add a "drop blocks"/throw exception when the operation is running from the main thread (I would have to implement it in the newest version)

In my opinion the proper approach would be to use solution 2. Never the less I'll have to add the solution 3 so no one else has the same problem.

**Comment by Postremus** [comment]
Thanks for your quick reply. I will try to implement the paste with the AWE API. I will reply to this issue once it is resolved, or when i have problems with using the api.

I have no idea how the block queue could be full. It is set to "queue-max-size: 10000000" in the config. our schematics are usually only around 100k blocks each.

**Comment by Postremus** [comment]
Got it working.
A few small things that i noticed: 
-When the paste is done in fast mode, the chunks are not updated - the player does not see the pasted schematic till he reconnects to the server.
-It takes a few seconds, till the paste even starts.
-It seems like the paste takes longer in general, but could be wrong on that.

-Is there a maven artifact of the api available somewhere? was not able to find it.
-Is there somewhere a already generated javadoc online available?

**Comment by SBPrime** [comment]
-**When the paste is done in fast mode, the chunks are not updated - the player does not see the pasted schematic till he reconnects to the server.** - I'm 99% sure that this is normal in worldedit fastmode

-**It takes a few seconds, till the paste even starts.** - I would need to see your code to say something specific. In my opinion the delay that you see is between start and flush.

-**Is there a maven artifact of the api available somewhere? was not able to find it.** - no artifact available. 

-**Is there somewhere a already generated javadoc online available?** - not yet. If i find a way to host it I'll publish it.

I would suggest 2 things:
- disable the WorldEdit queue mode (this way AWE starts placing blocks immediately)
- check the DirectChunk API its 100'ts times faster then the clasic WorldEdit API.

**Comment by Postremus** [comment]
-**I'm 99% sure that this is normal in worldedit fastmode** -I just tested it with we 6.1, "//fast", "//paste". The pasted thing appaered immediately.

-**check the DirectChunk API its 100'ts times faster then the clasic WorldEdit API.** -Is it available for 1.7.10? I read somewhere that it is only 1.8.3 (?).

-**I would need to see your code to say something specific. In my opinion the delay that you see is between start and flush.** -basicly the same code as the one you suggested in #9.
I was not able to call "AsyncWorldEditHandler.registerCompletion(player);", because it is not in the api, so i left it out.

**Comment by SBPrime** [comment]
-**I just tested it with we 6.1, "//fast", "//paste". The pasted thing appaered immediately.** - This might be an an error I'll look at this

-**Is it available for 1.7.10? I read somewhere that it is only 1.8.3 (?).** - If you need it for 1.7.10 (Is it why not 1.8? a moded server?) please send me the **EXACT** version of the server I'll see what I can do.

-**basicly the same code as the one you suggested in #9.
I was not able to call "AsyncWorldEditHandler.registerCompletion(player);", because it is not in the api, so i left it out.** - What version did you use? From the 1'st or the 2'nd post? "editSession.enableQueue();" this might be responsible for the small delay that you notice.

**Comment by Postremus** [comment]
-**If you need it for 1.7.10 (Is it why not 1.8? a moded server?)** - We let our community decide, if we should upgrade to 1.8.x. The result a few months back was, that more users still wanted to play 1.7.10 than 1.8.x.

-**please send me the EXACT version of the server I'll see what I can do.**
[06:47:46 INFO]: This server is running CraftBukkit version git-Spigot-b801989 (MC: 1.7.10) (Implementing API version 1.7.10-R0.1-SNAPSHOT)

-**What version did you use? From the 1'st or the 2'nd post?** -Removed the call to enableQueue, the past now starts without delay.
my code now: https://gist.github.com/Postremus/3087b9905a92af23afdd

**Comment by SBPrime** [comment]
Reported bug fixed in 3.2.1.

Regarding the support for 1.7.10-R0.1 I moved it to a new issue (#20).
